engine/src/main/java/org/terasology/i18n/TranslationSystemImpl.java

```diff
@@ -90,31 +77,72 @@ public TranslationProject getProject(Uri name) {
-                    return opt.get();
-                } else {
-                    logger.warn("No translation for '{}'", id);
-                    return "?" + uri.getFragmentName() + "?";
-                }
+    public String translate(String text, Locale otherLocale, Object... arguments) {
+        Object[] expandedArguments = expandFormatArguments(otherLocale, arguments);
+        TemplateEngine templateEngine = new TemplateEngineImpl(id ->
+                translateFragment(otherLocale, id, expandedArguments));
+
+        return templateEngine.transform(text);
+    }
+
+    /**
+     * Expand string arguments as if they are translatable templates.
+     *
+     * Leaves non-string arguments as they are.
+     * <p>
+     * One level of expansion only; String arguments may not require arguments of their own.
+     * <p>
+     * Example:
+     * <pre>{@code
+     *     Object[] arguments = {17, "${engine:menu#credits}", new Date(16e11)};
+     *     Object[] expandedArguments = expandFormatArguments(locale, arguments);
+     *
+     *     println(expandedArguments)  // {17, "Credits", Date(16e11)}
+     * }</pre>
+     *
+     * @param otherLocale FIXME: why <em>is</em> this "other" locale and not just "locale"?
+     * @param arguments arguments for {@link java.text.MessageFormat}
+     * @return argument array with replaced Strings
+     */
+    private Object[] expandFormatArguments(Locale otherLocale, Object[] arguments) {
+        Object[] expandedArguments = new Object[arguments.length];
+
+        // This engine is for template-processing each individual argument.
+        // We don't have nested argument providers; these get no argument list.
+        TemplateEngine templateEngine = new TemplateEngineImpl(id ->
+                translateFragment(otherLocale, id));
+
+        for (int i = 0; i < arguments.length; i++) {
+            if (arguments[i] instanceof String) {
+                expandedArguments[i] = templateEngine.transform((String) arguments[i]);
             } else {
-                logger.warn("Invalid project id '{}'", id);
-                return "?" + uri.getFragmentName() + "?";
+                expandedArguments[i] = arguments[i];
             }
-        });
+        }
+        return expandedArguments;
+    }
 
-        return templateEngine.transform(text);
+    private String translateFragment(Locale otherLocale, String id, Object... arguments) {
+        ResourceUrn uri = new ResourceUrn(id);
+        SimpleUri projectUri = new SimpleUri(uri.getModuleName(), uri.getResourceName());
+        TranslationProject project = getProject(projectUri);
+        if (project != null) {
+            Optional<String> opt = project.translate(uri.getFragmentName(), otherLocale, arguments);
+            if (opt.isPresent()) {
+                return opt.get();
+            } else {
+                logger.warn("No translation for '{}'", id);
+                return "?" + uri.getFragmentName() + "?";
+            }
```

Comment:
gnarf, this usage of `Optional` is kind of opposite to idiomatic 🧐  Waiting for [`ifPresentOrElse`](https://docs.oracle.com/javase/9/docs/api/java/util/Optional.html#ifPresentOrElse-java.util.function.Consumer-java.lang.Runnable-) to be available in the engine...

What do you think about the following?

```suggestion
            Optional<String> maybeTranslation= project.translate(uri.getFragmentName(), otherLocale, arguments);
            if (!maybeTranslation.isPresent()) {
                logger.warn("No translation for '{}'", id);
            }
            return maybeTranslation.orElse("?" + uri.getFragmentName() + "?");
```

Reply:
That is another way to write it. The variable name is better, but aside from that it doesn't add any clarity for me. Using an if-statement for the logger and then `.orElse` for the return value is two ways of testing for the same condition, which is unappealing.

By its signature, this method has to unwrap the Option somehow, I don't know if it has anything to gain from trying to do functional composition.

Reply:
Yeah, I know 😔 I'd like to chain somewhat like this:

```java
project.translate(uri.getFragmentName(), otherLocale, arguments)
  .ifEmpty(() -> logger.warn("No translation for '{}'", id))
  .orElse("?" + uri.getFragmentName() + "?");
```

But I agree, this `if-else` is probably a better solution here for the sake of readability...

... oh, wait:

```suggestion
            project.translate(uri.getFragmentName(), otherLocale, arguments)
            		.orElseGet(() -> {
            				logger.warn("No translation for '{}'", id);
							return "?" + uri.getFragmentName() + "?";
            		});
```